From a4285a8b2b3771608006ca01c65a475dcb4c5500 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Fri, 17 Sep 2021 16:29:48 +0800
Subject: [PATCH] drivers: llce-mailbox: change GFP_KERNEL to GFP_ATOMIC to fix
 debug calltrace

The llce_logger driver will show the following calltrace when debug
configures enabled:

 BUG: sleeping function called from invalid context at mm/slab.h:513
 in_atomic(): 1, irqs_disabled(): 128, non_block: 0, pid: 338, name: dhcpcd
 
 ......

 CPU: 1 PID: 338 Comm: dhcpcd Not tainted 5.10.63-yocto-standard #1
 Hardware name: Freescale S32G274 (DT)
 Call trace:
 dump_backtrace+0x0/0x1d4
 show_stack+0x24/0x30
 dump_stack+0xf0/0x13c
 ___might_sleep+0x19c/0x244
 __might_sleep+0x60/0x9c
 kmem_cache_alloc_trace+0x318/0x47c
 llce_logger_startup+0x5c/0xa0 [llce_mailbox]
 llce_mb_startup+0x3c/0x50 [llce_mailbox]
 mbox_request_channel+0x21c/0x2b4
 llce_logger_open+0x7c/0xf4 [llce_logger]
 __dev_open+0x128/0x1f4
 __dev_change_flags+0x1d4/0x254
 dev_change_flags+0x30/0x70
 devinet_ioctl+0x68c/0x700
 inet_ioctl+0x1e0/0x254
 sock_do_ioctl+0x50/0x1c4
 sock_ioctl+0x1fc/0x49c
 __arm64_sys_ioctl+0xb4/0x100
 el0_svc_common.constprop.0+0x9c/0x1a0
 do_el0_svc+0x78/0xa0
 el0_svc+0x20/0x30
 el0_sync_handler+0x1a4/0x1b0
 el0_sync+0x180/0x1c0

This is due to GPF_KERNEL flag used when memory allocation in llce_logger_startup() with spin lock
acquired. This patch will change the GPF_KERNEL to GPF_ATOMIC to fix the issue.

Upstream-Status: Pending

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/mailbox/llce-mailbox.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/mailbox/llce-mailbox.c b/drivers/mailbox/llce-mailbox.c
index 3c6299df8c3e..2e2b9c9fac56 100644
--- a/drivers/mailbox/llce-mailbox.c
+++ b/drivers/mailbox/llce-mailbox.c
@@ -869,7 +869,7 @@ static int llce_logger_startup(struct mbox_chan *chan)
 	priv->state = LLCE_REGISTERED_CHAN;
 
 	/* Make room for POP leftovers */
-	priv->data = kmalloc(sizeof(struct llce_logger_data), GFP_KERNEL);
+	priv->data = kmalloc(sizeof(struct llce_logger_data), GFP_ATOMIC);
 	if (!priv->data)
 		ret = -ENOMEM;
 
-- 
2.25.1

